### ISSUE TITLE
High convergence deltas in `deep_lift_shap` for specific input-reference pairs

### ISSUE THREAD

**Issue by caenrigen** [jmschrei/tangermeme#48]
Hey Jacob,

I have been training some CNNs and run into several warnings during the `deep_lift_shap` but only for specific input-reference pairs. I am new to DeepLIFT/SHAP methods. Is this expected to happen fairly often?

I spent a day debugging this and gathered some info. I can share a full example later (model weights + data), here is the gist:

I managed to skim the model down to only this:

```python
model = nn.Sequential(
        nn.Conv1d(4, 20, kernel_size=15, padding=7, bias=False),
        nn.MaxPool1d(kernel_size=15, stride=15),
        nn.Conv1d(20, 20, kernel_size=5, padding=2, bias=False),
        nn.MaxPool1d(kernel_size=5, stride=5),
        nn.Flatten(),
        nn.Linear(260, 1),
    )
```

For a specific X and (dinuc-shuffled) reference I get these deltas (last value = `3.2573e-03`):

```python
[1.4901e-08, 8.9407e-08, 1.4901e-08, 7.4506e-08, 2.2352e-08, 1.1548e-07,
        3.7253e-09, ..., 7.4506e-08, 1.0431e-07, 3.2573e-03]
```

The model outputs for this case were: `model(X) = 1.4073`, `model(ref) = 1.3880`

I have run into several examples by now, depending on the model higher deltas can happen as well (e.g. `~0.01`), but the rest of the references in the same batch are always fine.

Some interesting observations:

1. I did not test extensively but I did not run into this issue, so far, if I train the model with any of the 2 `nn.MaxPool1d` removed.
2. I am able to find examples of {X, reference}-pairs when running `deep_lift_shap` on either `device="cpu"` or `device="mps"` (Apple Silicon built-in GPU), BUT:
3. The {X, reference}-pairs for which the deltas are high seems to be device-dependent, i.e. `deep_lift_shap(sample_a, device="cpu") --> high delta` but `deep_lift_shap(sample_a, device="mps") --> all good`. And vice-versa, I have an example of `deep_lift_shap(sample_b, device="mps") --> high delta` but `deep_lift_shap(sample_b, device="cpu") --> all good`.

I played around with the epsilon in `_maxpool` hook but it did not solve it.

@jmschrei have you encountered this problem before? Is it safe to ignore? I am wondering If there might be some deeper issues somewhere

**Comment by dbogdano** [comment]
Hello. Any insight into why this issue? I too am experiencing high convergence deltas when running deep_lift_shap with a fairly simple model, similar to @caenrigen 's simplified model in their first post, containing MaxPool1d layers. Removing both maxpool layers from the model fixes the issue, but reporting their existance with `additional_nonlinear_ops = {nn.MaxPool1d:_maxpool}` does not. 

Removing other nonlinear layers from the model or running deep_lift_shap at 64-bit precision does not affect this issue as well. 

The issue appears to persist for all input-reference pairs, so it may be a separate problem than the one at hand in this thread, though I think the cause for both issues may stem from MaxPool1D.

**Comment by jmschrei** [comment]
This turned out to be a real bug in tangermeme, and your reproduction zip is what made it findable — thanks for putting that together. Fixed in #95.

The max-pool rule carries one multiplier between the observed and the reference half of the batch, and a pooling layer further up reads both halves. Where an input does not change at all between example and reference the rescale ratio has no meaning, so the rule substituted `grad_input` there — the ordinary gradient, which *differs* between the two halves. Your second pool then read two different values and stopped summing to delta.

That accounts for the properties you reported that never fitted a precision explanation:

- **Only some input-reference pairs.** It needs a position where the two happen to coincide exactly.
- **Device-dependent.** Which positions coincide depends on the arithmetic.
- **Survives fp64.** It is an algorithmic error, not a rounding one.
- **Removing either pool fixes it.** It takes two pools — one to write the mismatched value, one to read it.
- **Changing the epsilon in `_maxpool` does not help.** You were right about that. The epsilon moves which positions are affected, not whether the halves disagree. It is gone now: only an exact zero is guarded, since the routed change at a position is bounded by the change in that position and the ratio cannot diverge.

Running your checkpoint against the X/reference sets you attached, at fp64:

| pair | before | after |
|---|---|---|
| `cpu_108` | 1.81e-03 | 9.2e-16 |
| `cpu_117` | 3.26e-03 | 1.1e-15 |
| `mps_82`  | 9.3e-16  | 9.3e-16 |

`cpu_117` reproduces the `3.2573e-03` from your first post exactly. `mps_82` is the one case where Jacob's original diagnosis was right — it was already clean at fp64 before the fix, so that one really was fp32 noise.

Worth knowing if you use autocast: this is intermittent at fp32/fp64 but close to systematic in bf16/fp16, because coarse rounding makes an unchanged input common. `SharedPool` in fp16 improved by a factor of 60.

#95 also fixes a second, unrelated max-pool bug affecting overlapping windows (`kernel_size > stride`), which is #49. Your pools do not overlap, so that one never affected you.

**Comment by jmschrei** [comment]
^ claude

Closing this now, re-open if you still encounter issues.
